### PR TITLE
Rake task for managing plugins

### PR DESCRIPTION
**Pull request by jonasheinrich**
What we plan on doing:
- put steps into transactions
- include more sanity checks (e.g. a plugin is only listed once)
- introduce a blacklist-key for every plugin

What we might do:
- split the task up for developers
- rake plugin:add[community]

Comments welcome!

This is the relevant work package: https://community.openproject.org/work_packages/18720

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://jonasheinrich-feature-rake-tasks-for-administering-plugins.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://jonasheinrich-feature-rake-tasks-for-administering-plugins.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by jonasheinrich** [comment]
@opf/developers Comments welcome!

File: lib/tasks/plugins.rake [hunk 1]
```diff
@@ -88,3 +88,17 @@ namespace :redmine do
     end
   end
 end
+
+namespace :plugins do
+desc 'Adds Meeting plugin for now'
```
Comment (NobodysNightmare) [inline lib/tasks/plugins.rake]: Are the descriptions still valid?

I'll have to do further code lookup, but I hope that this will do more than add/remove the _Meeting_ plugin.

**Comment by jonasheinrich** [comment]
This is the relevant work package: https://community.openproject.org/work_packages/18720

File: lib/plugin_manager.rb [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+  end
+
+  def _bundle
+    Bundler.with_clean_env do
+      system 'bundle install --no-deployment'
+    end
+  end
+
+  def _migrate
+    system "RAILS_ENV='#{@environment}' rake db:migrate"
```
Comment (linki) [inline lib/plugin_manager.rb]: how about using `bundle exec` to run rake?

File: lib/plugin_manager.rb [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+  end
+end
+
+class Plugin
+  def self._available?(name)
+    _available_plugins[name]
+  end
+
+  def self._available_plugins
+    @available_plugins || _load_available_plugins
```
Comment (linki) [inline lib/plugin_manager.rb]: most people use `@available_plugins ||= _load_available_plugins`. you can drop the instance var assignment in `_load_available_plugins` then.

File: lib/plugin_manager.rb [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+      exit
+    end
+    @available_plugins = YAML.load_file(PLUGINS_YML_PATH)
+  end
+
+  attr_reader :name
+
+  def initialize(name)
+    if name == 'pdf-inspector'
+      @name = 'pdf-inspector'
```
Comment (linki) [inline lib/plugin_manager.rb]: could you please add a comment on why this is necessary and when it can be removed.

File: lib/plugin_manager.rb [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    else
+      unless Plugin._available?(name)
+        puts 'Could not find plugin, abort!'
+        exit
+      end
+      @name = name
+      @url = Plugin._available_plugins[name][:url]
```
Comment (linki) [inline lib/plugin_manager.rb]: either you're calling a private method here or it shouldn't begin with an underscore in the first place.. pick one :)

File: lib/plugin_manager.rb [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+      @branch = Plugin._available_plugins[name][:branch]
+    end
+  end
+
+  def _not_needed_by_any_other_than?(plugin)
```
Comment (linki) [inline lib/plugin_manager.rb]: :+1: @houndci

Comment (linki) [inline lib/plugin_manager.rb]: or maybe not... `plugin` is used in the method ^^

**Comment by linki** [comment]
please add tests :racehorse:

**Comment by ulferts** [comment]
@jonasheinrich this is labeled 'WIP' is that correct?

**Comment by kgalli** [comment]
@ulferts Yes it still is. I suggested to change the task to use plugins only available on rubygems for now. I will update the plugins/gems on rubygems.org and adjust the rake task accordingly. Otherwise we will have a lot of trouble regarding dependencies of plugins.

**Comment by linki** [comment]
**tl/dr** check in `Gemfile.lock`, `schema.rb`, etc. into the core and config repos. do not deal with any specific plugins in the core code.

# 

please let us try to not intermix openproject's core and openproject-plugin-combinations (configs, in finnlabs speak) with each other any further.

for example: the core has its set of gems, therefore it has its very own `Gemfile.lock`. the core also has its clearly defined database schema, resulting from the set of migrations at the moment, therefore it has its very own `schema.rb`, which **should** be checked it.

configs contain plugins, which contain gem dependencies, therefore configs have their own `Gemfile.lock` which differs from the one of the core and should be checked into the `config`s repo. plugins also change the database by adding migrations, therefore the database schema differs from the core's. logically, there will be a different `schema.rb` which **should** be checked into the config's repo as well.

here, we have a rake task that deals with a file called `Gemfile.plugins`, which only exists in configs. it also references some specific plugins by name. therefore, the rake task as it is now should not live in the core (it could be callable from the project root by some means, though).

**Comment by myabc** [comment]
@jonasheinrich this is targeting 4.2, right? (I set the milestone)

**Comment by kgalli** [comment]
@myabc Yes would be great to see this in 4.2. I suggest we publish all plugins as gems on rubygems.org so this rake task can be implemented as easy as possible. At least in regard to resolving dependencies. I already removed the relatively complex dependency logic but it is still in progress.

**Comment by jonasheinrich** [comment]
I'm gonna close this since it is not and won't be finished in the next time. Again, here is the relevant work package: https://community.openproject.org/work_packages/18720
